crates/jstz_core/src/kv/transaction.rs

```diff
@@ -43,258 +40,353 @@ use std::rc::Rc;
 /// +---------------+
 /// ```
 ///
-/// Current implementation does NOT support concurrent transactions.
-
-#[must_use]
-#[derive(Debug)]
-pub struct Transaction {
-    parent: Option<Rc<RefCell<Transaction>>>,
-    remove_set: BTreeSet<OwnedPath>,
-    snapshot: Snapshot,
-}
+/// NOTE: Current implementation does NOT support concurrent transactions
 
-impl Finalize for Transaction {}
+/// A key is a path in durable storage
+pub type Key = OwnedPath;
 
-unsafe impl Trace for Transaction {
-    empty_trace!();
-}
+// A lookup map is a history of edits of a given key in order of least-recent to most-recent
+// This allows O(log n) lookups, and O(log n) commits / rollbacks (amortized by # of inserts / removals).
```

Comment:
To lookup the snapshot with the most recent version of the key (comment explains this + the performance reasoning behind it)

Reply:
I might be missing something here, but it seems to me that if you need to lookup just the last version it is enough to have just stack. As I understand it, you never really lookup something in the middle of the history of edits, just the last value as the invariant is that if you want to lookup something earlier, you have already popped later edits from the history.

Reply:
No. This is not the case. It is possible to read values from previous snapshots in the stack (provided there has been no mutable changes to the value)

Reply:
Can you show somewhere in the code where exactly can it happen so that I can better understand it? If there were no mutable changes to the value, you still care just about the last change of the value.

Reply:
You can see this behaviour in `Transaction::lookup` method. Namely note that it obtains the most recent snapshot index of the key and they lookups the value in the snapshot.

Reply:
> If there were no mutable changes to the value, you still care just about the last change of the value.

We care about the most recent version of the value, that does not imply that the value is in the most recent snapshot.

Reply:
ah yes, makes sense - I was confused thinking that it logarithmic based on the length of the history based on the comment.